lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -913,57 +924,83 @@ selectSeqStatePendingIxs ssid =
+            [ RndStateCheckpointWalletId ==. wid
             , RndStateCheckpointSlot ==. sl
             ] []
-        let (rsid, RndState _ _ accIx gen) = (entityKey st, entityVal st)
-        addresses <- lift $ selectRndStateAddresses rsid False
-        pendingAddresses <- lift $ selectRndStateAddresses rsid True
+        addresses <- lift $ selectRndStateAddresses wid sl
+        pendingAddresses <- lift $ selectRndStatePending (entityKey stc)
         rndKey <- lift $ selectRndStateKey wid
         pure $ Rnd.RndState
             { rndKey = rndKey
-            , accountIndex = W.Index accIx
+            , accountIndex = W.Index (rndStateAccountIndex (entityVal st))
             , addresses = addresses
             , pendingAddresses = pendingAddresses
-            , gen = gen
+            , gen = rndStateCheckpointGen (entityVal stc)
             }
 
     deleteState wid = do
-        rsid <- fmap entityKey <$> selectList [ RndStateWalletId ==. wid ] []
-        deleteWhere [ RndStateAddressRndStateId <-. rsid ]
+        cpid <- fmap entityKey <$> selectList [ RndStateCheckpointWalletId ==. wid ] []
+        deleteWhere [ RndStatePendingAddressCheckpointId <-. cpid ]
+        deleteWhere [ RndStateCheckpointWalletId ==. wid ]
+        deleteWhere [ RndStateAddressWalletId ==. wid ]
         deleteWhere [ RndStateWalletId ==. wid ]
 
 insertRndStateAddresses
-    :: RndStateId
-    -> Bool
+    :: W.WalletId
+    -> W.SlotId
+    -> Map (W.Index 'W.Hardened 'W.AccountK, W.Index 'W.Hardened 'W.AddressK) W.Address
+    -> SqlPersistT IO ()
+insertRndStateAddresses wid sl addresses =
+    void $ dbChunked insertMany_
+        [ RndStateAddress wid sl accIx addrIx addr
+        | ((W.Index accIx, W.Index addrIx), addr) <- Map.assocs addresses
+        ]
+
+insertRndStatePending
+    :: RndStateCheckpointId
     -> Map (W.Index 'W.Hardened 'W.AccountK, W.Index 'W.Hardened 'W.AddressK) W.Address
     -> SqlPersistT IO ()
-insertRndStateAddresses rsid pending addresses =
+insertRndStatePending cpid addresses =
     void $ dbChunked insertMany_
-        [ RndStateAddress rsid accIx addrIx addr pending
+        [ RndStatePendingAddress cpid accIx addrIx addr
         | ((W.Index accIx, W.Index addrIx), addr) <- Map.assocs addresses
         ]
 
 selectRndStateAddresses
-    :: RndStateId
-    -> Bool
+    :: W.WalletId
+    -> W.SlotId
+    -> SqlPersistT IO (Map (W.Index 'W.Hardened 'W.AccountK, W.Index 'W.Hardened 'W.AddressK) W.Address)
+selectRndStateAddresses wid sl = do
+    addrs <- fmap entityVal <$> selectList
+        [ RndStateAddressWalletId ==. wid
+        , RndStateAddressSlot ==. sl
+        ] []
+    pure $ Map.fromList $ map assocFromEntity addrs
+  where
+    assocFromEntity (RndStateAddress _ _ accIx addrIx addr) =
+        ((W.Index accIx, W.Index addrIx), addr)
+
+selectRndStatePending
+    :: RndStateCheckpointId
     -> SqlPersistT IO (Map (W.Index 'W.Hardened 'W.AccountK, W.Index 'W.Hardened 'W.AddressK) W.Address)
```

Comment:
We should probably also fix this line, perhaps as suggested [here](https://github.com/input-output-hk/cardano-wallet/pull/672/files#r320179776).